### PR TITLE
c8d/list: Fix race condition when traversing containers

### PR DESCRIPTION
**Pull request by vvoland**
**- What I did**
Fix a panic spotted in https://github.com/moby/moby/pull/47983#issuecomment-2306706264.

Shared state is mutated inside a function which is called asynchronously:
https://github.com/moby/moby/blob/4f0d95fa6ee7f865597c03b9e63702cdcb0f7067/container/memory_store.go#L69-L83

Unfortunately, it looks like the race detector doesn't pick this up for some reason.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Hmm. So wondering; should synchronization for this happen in the memory store itself? I see it has locking, but (iiuc) for the store itself, and it looks like a RWMutex taking a read lock. However, the slice returned does not dereference the container; is the problem in that area? Should it either return a copy of the container, or should locking happen for each container? 😬 https://github.com/moby/moby/blob/d8f079deb6d9faf6540b756baec42ede172cd53e/container/memory_store.go#L86

File: daemon/containerd/image_list.go [hunk 1]
```diff
@@ -225,6 +225,8 @@ func (i *ImageService) imageSummary(ctx context.Context, img images.Image, platf
 	var best *ImageManifest
 	var bestPlatform ocispec.Platform
 
+	var mut sync.Mutex
```
Comment (laurazard) [inline daemon/containerd/image_list.go:228]: Ah, I was going to say "why don't we move it inside `i.WalkReachableIamgeManifests`/next to `msftSummary` but I see it's also guarding `countainersCount`.

Can you move this closer to `containersCount` and add a comment explaining that it guards access to `containersCount` and `msftSummary`?

Comment (vvoland) [reply daemon/containerd/image_list.go:229]: Done

**Comment by vvoland** [comment]
@thaJeztah - There's no need to lock the container object itself, if the caller doesn't really mutate the container (like we do here), so it wouldn't make sense to lock inside `ApplyAll`.
We only read the `ID` and `ImageManifest` property of the image, which don't change after the container is created, so it's fine to read from the reference in this case.

**Comment by vvoland** [comment]
The only reason why this sneaked past me is:

- ApplyAll doesn't look like it would act asynchronously (and I didn't bother to check the implementation 🙈)
- For some reason, it wasn't detected when I tested this with `-race`. Not sure why race detectors doesn't catch this, will need to investigate it more.

**Review by laurazard — APPROVED** [review]
LGTM

File: daemon/containerd/image_list.go [hunk 2]
```diff
@@ -355,6 +358,11 @@ func (i *ImageService) imageSummary(ctx context.Context, img images.Image, platf
 
 		i.containers.ApplyAll(func(c *container.Container) {
 			if c.ImageManifest != nil && c.ImageManifest.Digest == target.Digest {
+				// ApplyAll calls the function asynchronously for each
+				// container in a separate goroutine, so we need to sync
+				// mutations to the summary.
+				mut.Lock()
+				defer mut.Unlock()
 				mfstSummary.ImageData.Containers = append(mfstSummary.ImageData.Containers, c.ID)
```
Comment (thaJeztah) [inline daemon/containerd/image_list.go:366]: slightly wondering; now that we have the locking.. are we effectively defeating the whole purpose of the `ApplyAll()` logic to use a goroutine? Given that the code in the goroutine is .. basically “nothing” (check the digest and append), Would it make sense to instead just use a regular loop? just loop over `container.List()`; https://github.com/moby/moby/blob/d8f079deb6d9faf6540b756baec42ede172cd53e/container/memory_store.go#L44-L50

Only downside from that one is that it also sorts the containers (which we don’t need I guess)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by thaJeztah** [comment]
Ah, dang; linting failure somewhere 

```
daemon/containerd/image_list_test.go:7: File is not `goimports`-ed (goimports)
	"github.com/docker/docker/container"

**Comment by thaJeztah** [comment]
I fixed teh linting issue; problem was in the imports;

```patch
diff --git a/daemon/containerd/image_list_test.go b/daemon/containerd/image_list_test.go
index 50ba317e03..28b61db4b5 100644
--- a/daemon/containerd/image_list_test.go
+++ b/daemon/containerd/image_list_test.go
@@ -4,7 +4,6 @@ import (
        "context"
        "errors"
        "fmt"
-       "github.com/docker/docker/container"
        "math/rand"
        "os"
        "path/filepath"
@@ -23,6 +22,7 @@ import (
        "github.com/containerd/log/logtest"
        "github.com/containerd/platforms"
        imagetypes "github.com/docker/docker/api/types/image"
+       "github.com/docker/docker/container"
        daemonevents "github.com/docker/docker/daemon/events"
        "github.com/docker/docker/internal/testutils/specialimage"
        "github.com/opencontainers/go-digest"
```

**Pull request merged**